### ISSUE TITLE
zfs-mount.service broken after upgrade 0.6.3 -> 0.6.4

### ISSUE THREAD

**Issue by Uglymotha** [zfsonlinux/zfs#3311]
After upgrade to 0.6.4 zfs filesystems are no linger mounted by the zfs-mount.service systemd unit, yielding these errors:
Apr 17 09:39:59 san zfs[2671]: mount: mount(2) failed: No such file or directory
Apr 17 09:39:59 san zfs[2671]: cannot mount 'bla': No such device or address

Running 'zfs mount -a' as root from the shell works fine. Looks a bit like old bug #115

[root@san ~]# zpool list
NAME   SIZE  ALLOC   FREE  EXPANDSZ   FRAG    CAP  DEDUP  HEALTH  ALTROOT
bla    232G  77.5K   232G         -     0%     0%  1.00x  ONLINE  -

[root@san ~]# zfs list
NAME   USED  AVAIL  REFER  MOUNTPOINT
bla   65.5K   228G    19K  /bla

[root@san ~]# systemctl start zfs-mount
Job for zfs-mount.service failed. See "systemctl status zfs-mount.service" and "journalctl -xe" for details.

[root@san ~]# systemctl status zfs-mount
● zfs-mount.service - Mount ZFS filesystems
   Loaded: loaded (/lib/systemd/system/zfs-mount.service; static; vendor preset: enabled)
   Active: failed (Result: exit-code) since Fri 2015-04-17 09:39:59 CEST; 17s ago
  Process: 2671 ExecStart=/usr/sbin/zfs mount -a (code=exited, status=1/FAILURE)
Main PID: 2671 (code=exited, status=1/FAILURE)

Apr 17 09:39:59 san systemd[1]: Starting Mount ZFS filesystems...
Apr 17 09:39:59 san zfs[2671]: mount: mount(2) failed: No such file or directory
Apr 17 09:39:59 san zfs[2671]: cannot mount 'bla': No such device or address
Apr 17 09:39:59 san systemd[1]: zfs-mount.service: main process exited, code=exited, status=1/FAILURE
Apr 17 09:39:59 san systemd[1]: Failed to start Mount ZFS filesystems.
Apr 17 09:39:59 san systemd[1]: Unit zfs-mount.service entered failed state.
Apr 17 09:39:59 san systemd[1]: zfs-mount.service failed.

[root@san ~]# zfs mount -a

[root@san ~]# df
Filesystem     1K-blocks   Used Available Use% Mounted on
/dev/root        1996624 425716   1451436  23% /
devtmpfs         1016308      0   1016308   0% /dev
tmpfs            1018512      0   1018512   0% /dev/shm
tmpfs            1018512   1152   1017360   1% /run
tmpfs            1018512      0   1018512   0% /sys/fs/cgroup
tmpfs            1018512      0   1018512   0% /tmp
bla            239468416      0 239468416   0% /bla

This issue was reproduced with kernels 3.14.38 and 3.18.11.

**Comment by FransUrbo** [comment]
Please read http://zfsonlinux.org/faq.html#HowDoIReportProblems.

**Comment by Uglymotha** [comment]
So I straced a regular zfs mount -a executed from the shell and one executed by system.

zfs mount -a forks this process:
execve("/bin/mount", ["/bin/mount", "-t", "zfs", "-o", "defaults,atime,dev,exec,rw,suid,"..., "test", "/test/bla"], [/* 5 vars */])
 = 0

Eventually during the mount system call of this spawned process a difference in behavior is shown.

Executing zfs mount -a from the cli the system call looks like:
[pid  3810] mount("test", "/test/bla", "zfs", MS_MGC_VAL, "xattr,zfsutil") = 0

Excuted by system it's:
mount("/test", "/test/bla", "zfs", MS_MGC_VAL, "xattr,zfsutil") = -1 ENOENT (No such file or directory)

**Comment by Uglymotha** [comment]
Could not give you all required info this week-end since I reverted back to a 0.6.3 build to test and verify. 

So here we go:

I am not using a distribution, compiling my own system from source (like Linux from scratch). Using the vanilla sources for all components. Currently only testing, but eventually system wwill be used for iSCSI / SMB. I reproduced and verified the issue by creating 4 builds. All builds were exactly the same except for kernel and zfs versions. Build A kernel 3.14.38 zfs 0.6.3. build B kernel 3.18.11 zfs 0.6.3. Build C kernel 3.14.38 zfs 0.6.4. Build D kernel 3.18.11 zfs 0.6.4. Builds A and B are fine, builds C and D suffer this issue. Conclusion here is it's ZFS.

Testing was also done on 2 different systems, issue reproduced on both with builds C and D. Conclusion here is it's still ZFS, not a hardware / VM issue.
1 Virtual (ESX, 1vCPU, 2Gib no ECC), 1 Hardware (AMD Athlon 64 x2, nForce 513, 2GiB no ECC).

The following output is all based on build C, if you like to see the output of other builds, let me know.

[root@san ~]# uname -a
Linux san 3.14.38-uglylinux-3.0 #1 SMP Sun Apr 19 13:27:35 CEST 2015 x86_64 GNU/Linux

root@san ~]# dmesg | grep -E 'SPL:|ZFS:'
[    1.110494] SPL: Loaded module v0.6.4-1
[    1.113474] ZFS: Loaded module v0.6.4-1, ZFS pool version 5000, ZFS filesystem version 5
[    9.062750] kernel[3064]: [    1.110494] SPL: Loaded module v0.6.4-1
[    9.063198] kernel[3064]: [    1.113474] ZFS: Loaded module v0.6.4-1, ZFS pool version 5000, ZFS filesystem version 5
[    9.185913] SPL: using hostid 0xaed9e424
[    9.190584] kernel[3064]: [    9.185913] SPL: using hostid 0xaed9e424

The following seems to be overkill in this case, the strace snippets above is a clear indication of what's the issue.
spl_hostid                                        2933515300
spl_hostid_path                                   /etc/hostid
spl_kmem_alloc_max                                2097152
spl_kmem_alloc_warn                               32768
spl_kmem_cache_expire                             2
spl_kmem_cache_kmem_limit                         2048
spl_kmem_cache_kmem_threads                       4
spl_kmem_cache_magazine_size                      0
spl_kmem_cache_max_size                           32
spl_kmem_cache_obj_per_slab                       8
spl_kmem_cache_obj_per_slab_min                   1
spl_kmem_cache_reclaim                            0
spl_kmem_cache_slab_limit                         16384
spl_taskq_thread_bind                             0
l2arc_feed_again                                  1
l2arc_feed_min_ms                                 200
l2arc_feed_secs                                   1
l2arc_headroom                                    2
l2arc_headroom_boost                              200
l2arc_nocompress                                  0
l2arc_noprefetch                                  1
l2arc_norw                                        0
l2arc_write_boost                                 8388608
l2arc_write_max                                   8388608
metaslab_bias_enabled                             1
metaslab_debug_load                               0
metaslab_debug_unload                             0
metaslab_fragmentation_factor_enabled             1
metaslab_lba_weighting_enabled                    1
metaslab_preload_enabled                          1
metaslabs_per_vdev                                200
spa_asize_inflation                               24
spa_config_path                                   /etc/zfs/zpool.cache
spa_load_verify_data                              1
spa_load_verify_maxinflight                       10000
spa_load_verify_metadata                          1
zfetch_array_rd_sz                                1048576
zfetch_block_cap                                  256
zfetch_max_streams                                8
zfetch_min_sec_reap                               2
zfs_arc_average_blocksize                         8192
zfs_arc_grow_retry                                5
zfs_arc_max                                       0
zfs_arc_memory_throttle_disable                   1
zfs_arc_meta_adjust_restarts                      4096
zfs_arc_meta_limit                                0
zfs_arc_meta_prune                                10000
zfs_arc_min                                       0
zfs_arc_min_prefetch_lifespan                     100
zfs_arc_p_aggressive_disable                      1
zfs_arc_p_dampener_disable                        1
zfs_arc_shrink_shift                              5
zfs_autoimport_disable                            1
zfs_dbuf_state_index                              0
zfs_deadman_enabled                               1
zfs_deadman_synctime_ms                           1000000
zfs_dedup_prefetch                                0
zfs_delay_min_dirty_percent                       60
zfs_delay_scale                                   500000
zfs_dirty_data_max                                208260710
zfs_dirty_data_max_max                            520651776
zfs_dirty_data_max_max_percent                    25
zfs_dirty_data_max_percent                        10
zfs_dirty_data_sync                               67108864
zfs_disable_dup_eviction                          0
zfs_expire_snapshot                               300
zfs_flags                                         0
zfs_free_leak_on_eio                              0
zfs_free_max_blocks                               100000
zfs_free_min_time_ms                              1000
zfs_immediate_write_sz                            32768
zfs_mdcomp_disable                                0
zfs_metaslab_fragmentation_threshold              70
zfs_mg_fragmentation_threshold                    85
zfs_mg_noalloc_threshold                          0
zfs_no_scrub_io                                   0
zfs_no_scrub_prefetch                             0
zfs_nocacheflush                                  0
zfs_nopwrite_enabled                              1
zfs_pd_bytes_max                                  52428800
zfs_prefetch_disable                              0
zfs_read_chunk_size                               1048576
zfs_read_history                                  0
zfs_read_history_hits                             0
zfs_recover                                       0
zfs_resilver_delay                                2
zfs_resilver_min_time_ms                          3000
zfs_scan_idle                                     50
zfs_scan_min_time_ms                              1000
zfs_scrub_delay                                   4
zfs_send_corrupt_data                             0
zfs_sync_pass_deferred_free                       2
zfs_sync_pass_dont_compress                       5
zfs_sync_pass_rewrite                             2
zfs_top_maxinflight                               32
zfs_txg_history                                   0
zfs_txg_timeout                                   5
zfs_vdev_aggregation_limit                        131072
zfs_vdev_async_read_max_active                    3
zfs_vdev_async_read_min_active                    1
zfs_vdev_async_write_active_max_dirty_percent     60
zfs_vdev_async_write_active_min_dirty_percent     30
zfs_vdev_async_write_max_active                   10
zfs_vdev_async_write_min_active                   1
zfs_vdev_cache_bshift                             16
zfs_vdev_cache_max                                16384
zfs_vdev_cache_size                               0
zfs_vdev_max_active                               1000
zfs_vdev_mirror_switch_us                         10000
zfs_vdev_read_gap_limit                           32768
zfs_vdev_scheduler                                noop
zfs_vdev_scrub_max_active                         2
zfs_vdev_scrub_min_active                         1
zfs_vdev_sync_read_max_active                     10
zfs_vdev_sync_read_min_active                     10
zfs_vdev_sync_write_max_active                    10
zfs_vdev_sync_write_min_active                    10
zfs_vdev_write_gap_limit                          4096
zfs_zevent_cols                                   80
zfs_zevent_console                                0
zfs_zevent_len_max                                64
zil_replay_disable                                0
zil_slog_limit                                    1048576
zio_delay_max                                     30000
zio_requeue_io_start_cut_in_line                  1
zvol_inhibit_dev                                  0
zvol_major                                        230
zvol_max_discard_blocks                           16384
zvol_threads                                      32

**Comment by FransUrbo** [comment]
"It have been said" (but since I don't use systemd - I avoid it forcefully - I have not been able to verify) that the systemd files distributed by ZoL is "broken" (somehow). You can wait for someone to comment on this in this issue, or you can take it to the list. I recommend the latter.

**Comment by kkoning** [comment]
So... I had what seems like a similar problem.  I had noticed that, while I could mount everything fine as root, I would see a similar error if I tried to mount a volume from the root directory.  After doing a ton of fussing with different versions of the startup script, I added a cd /root (root's home directory) before doing the actual import.  Things started to work just fine after that.

I don't know if this is the same issue you are seeing, but looking at this might be worth a try.
